File: server/src/main/scala/producer/producer_task/ProducerTask.scala
package producer.producer_task

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import _root_.producer.message_generator.MessageGenerator
import _root_.producer.pulsar_producer_config.PulsarProducerConfig
import consumer.session_target.topic_selector.TopicSelector

case class ProducerTask(
    topicSelector: TopicSelector,
    messageGenerator: MessageGenerator,
    producerConfig: Option[PulsarProducerConfig],
    numMessages: Option[Long],
    limitDurationNanos: Option[Long],
    intervalNanos: Option[Long]
)

object ProducerTask:
    def fromPb(v: pb.ProducerTask): ProducerTask =
        ProducerTask(
            topicSelector = v.topicSelector.map(TopicSelector.fromPb).get,
            messageGenerator = v.messageGenerator.map(MessageGenerator.fromPb).get,
            producerConfig = v.producerConfig.map(PulsarProducerConfig.fromPb),
            numMessages = v.numMessages,
            limitDurationNanos = v.limitDurationNanos,
            intervalNanos = v.intervalNanos
        )

    def toPb(v: ProducerTask): pb.ProducerTask =
        pb.ProducerTask(
            topicSelector = Some(TopicSelector.toPb(v.topicSelector)),
            messageGenerator = Some(MessageGenerator.toPb(v.messageGenerator)),
            producerConfig = v.producerConfig.map(PulsarProducerConfig.toPb),
            numMessages = v.numMessages,
            limitDurationNanos = v.limitDurationNanos,
            intervalNanos = v.intervalNanos
        )

File: ui/components/ui/ProducerSession/producer-task/producer-task.ts
import * as pb from '../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { Int64Value } from 'google-protobuf/google/protobuf/wrappers_pb';
import { MessageGenerator, messageGeneratorFromPb, messageGeneratorToPb } from './message-generator/message-generator';
import { PulsarProducerConfig, pulsarProducerConfigFromPb, pulsarProducerConfigToPb } from './pulsar-producer-config/pulsar-producer-config';
import { TopicSelector } from '../../ConsumerSession/topic-selector/topic-selector';
import { topicSelectorFromPb, topicSelectorToPb } from '../../ConsumerSession/conversions/conversions';

export type ProducerTask = {
  type: 'producer-task',
  topicSelector: TopicSelector,
  messageGenerator: MessageGenerator,
  producerConfig: PulsarProducerConfig | undefined,
  numMessages: number | undefined,
  limitDurationNanos: number | undefined,
  intervalNanos: number | undefined,
};

export function producerTaskFromPb(v: pb.ProducerTask): ProducerTask {
  return {
    type: 'producer-task',
    topicSelector: topicSelectorFromPb(v.getTopicSelector()!),
    messageGenerator: messageGeneratorFromPb(v.getMessageGenerator()!),
    producerConfig: pulsarProducerConfigFromPb(v.getProducerConfig()!),
    numMessages: v.getNumMessages()?.getValue(),
    limitDurationNanos: v.getLimitDurationNanos()?.getValue(),
    intervalNanos: v.getIntervalNanos()?.getValue(),
  };
}

export function producerTaskToPb(v: ProducerTask): pb.ProducerTask {
  const producerTaskPb = new pb.ProducerTask();
  producerTaskPb.setTopicSelector(topicSelectorToPb(v.topicSelector));
  producerTaskPb.setMessageGenerator(messageGeneratorToPb(v.messageGenerator));

  if (v.producerConfig !== undefined) {
    producerTaskPb.setProducerConfig(pulsarProducerConfigToPb(v.producerConfig));
  }

  if (v.numMessages !== undefined) {
    producerTaskPb.setNumMessages(new Int64Value().setValue(v.numMessages));
  }
  if (v.limitDurationNanos !== undefined) {
    producerTaskPb.setLimitDurationNanos(new Int64Value().setValue(v.limitDurationNanos));
  }
  if (v.intervalNanos !== undefined) {
    producerTaskPb.setIntervalNanos(new Int64Value().setValue(v.intervalNanos));
  }

  return producerTaskPb;
}
